File: backend/controllers/restaurant_controller.py
from services.google_places import get_nearby_restaurants
from services.database import get_restaurant_tags, store_restaurant_tags
from services.groq_api import generate_restaurant_tags

def get_restaurants(lat, long, radius):
    """
    Main controller function to get restaurants with tags.
    
    Args:
        lat (float): Latitude
        long (float): Longitude
        radius (int): Search radius in meters
    
    Returns:
        list: List of restaurant objects with tags
    """
    result = []
    
    # Get nearby restaurants from Google Places API
    restaurants = get_nearby_restaurants(lat, long, radius)
    
    # Process each restaurant
    for restaurant in restaurants:
        restaurant_name = restaurant.get('name')
        
        # Try to get tags from database
        tags = get_restaurant_tags(restaurant_name)
        
        # If tags don't exist, generate them using Groq
        if not tags:
            # Generate tags using AI
            tags = generate_restaurant_tags(restaurant)
            
            # Store the restaurant and tags in database
            store_restaurant_tags(restaurant_name, tags)
        
        # Extract location data
        location = restaurant.get('geometry', {}).get('location', {})
        
        # Add restaurant with tags to result
        restaurant_data = {
            'name': restaurant_name,
            'place_id': restaurant.get('place_id'),
            'address': restaurant.get('vicinity'),
            'lat': location.get('lat'),
            'long': location.get('lng'),  # Google Places API uses 'lng', we map it to 'long'
            'rating': restaurant.get('rating'),
            'user_ratings_total': restaurant.get('user_ratings_total'),
            'tags': tags
        }
        
        result.append(restaurant_data)
    
    return result

File: api/_lib/controllers/restaurantController.ts
import { getNearbyRestaurants } from '../services/googlePlaces';
import { getRestaurantFromDb, storeRestaurant } from '../services/database';
import { generateRestaurantTags } from '../services/groqApi';

export interface RestaurantResult {
  name: string;
  place_id: string;
  address: string;
  lat: number;
  long: number;
  rating?: number;
  total_ratings?: number;
  tags: string[];
}

export async function getRestaurants(lat: number, long: number, radius: number): Promise<RestaurantResult[]> {
  const restaurants = await getNearbyRestaurants(lat, long, radius);
  const results: RestaurantResult[] = [];

  for (const restaurant of restaurants) {
    const { name, vicinity, rating, user_ratings_total, place_id, geometry } = restaurant;
    const { lat: rLat, lng: rLng } = geometry.location;

    const dbRecord = await getRestaurantFromDb(name);
    let tags: string[];

    if (dbRecord?.tags?.length) {
      console.log(`[Controller] Cache hit for "${name}"`);
      tags = dbRecord.tags;
    } else {
      console.log(`[Controller] Generating tags for "${name}"...`);
      tags = await generateRestaurantTags(restaurant);
      console.log(`[Controller] Tags for "${name}":`, tags);
      await storeRestaurant({
        name,
        tags,
        address: vicinity,
        rating,
        total_ratings: user_ratings_total,
        lat: rLat,
        long: rLng,
        place_id,
      });
    }

    results.push({ name, place_id, address: vicinity, lat: rLat, long: rLng, rating, total_ratings: user_ratings_total, tags });
  }

  return results;
}
